File: CC++/sort/MergeSort.cpp
#include <iostream>
using namespace std;

void merge(int* arr, int begin, int mid, int end, int maxE)
{
    int i = begin, j = mid + 1,
        k = begin;
    
    while (i <= mid && j <= end)
    {
        if (arr[i] % maxE < arr[j] % maxE)
        {
            arr[k] += (arr[i] % maxE)*maxE;
            i++;
        }
        else
        {
            arr[k] += (arr[j] % maxE)*maxE;
            j++;
        }
        k++;
    }

    for (; i <= mid; i++, k++)
        arr[k] += (arr[i] % maxE)*maxE;
    for (; j <= end; j++, k++)
        arr[k] += (arr[j] % maxE)*maxE;
    
    for (int i = begin; i <= end; i++)
        arr[i] /= maxE;
}

// Recursive merge sort with extra parameter, naxele
void mergeSortRec(int arr[], int beg, int end, int maxele)
{
    if (beg < end)
    {
        int mid = (beg + end) / 2;
        mergeSortRec(arr, beg, mid, maxele);
        mergeSortRec(arr, mid + 1, end, maxele);
        merge(arr, beg, mid, end, maxele);
    }
}

// This functions finds max element and calls recursive
// merge sort.
void mergeSort(int arr[], int n)
{
    int maxE = arr[0];
    for (int i = 1; i < n; i++)
        if (maxE < arr[i]) maxE = arr[i];
    mergeSortRec(arr, 0, n - 1, maxE + 1);
}

int main()
{
    int arr[] = {999, 612, 589, 856, 56, 945, 243};
    int n = sizeof(arr) / sizeof(arr[0]);

    mergeSort(arr, n);

    cout << "Sorted array \n";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    return 0;
}

File: Java/sort/MergeSort.java
public class MergeSort {
    // to store 2 value a, b in one position in array 
    // use arr[i] = a + b*max   // max is greater than both of a and b
    // => a = arr[i] % max & b = arr[i] / max
    static void merge(int arr[], int begin, int end, int mid, int max) {
        int i = begin,
            j = mid + 1,
            k = begin;
            
        for (; i <= mid && j <= end; k++) {
            if (arr[i] % max < arr[j] % max) {
                arr[k] += (arr[i] % max)*max;
                i++;
            }
            else {
                arr[k] += (arr[j] % max)*max;
                j++;
            }
        }

        for (; i <= mid; i++, k++)
            arr[k] += (arr[i] % max)*max;
        
        for (; j <= end; j++, k++)
            arr[k] += (arr[j] % max)*max;
        
        for (i = begin; i <= end; i++)
            arr[i] /= max;
    }

    static void mergeSort(int arr[], int begin, int end, int max) {
        if (begin < end) {
            int mid = (begin + end)/2;

            mergeSort(arr, begin, mid, max);
            mergeSort(arr, mid + 1, end, max);

            merge(arr, begin, end, mid, max);
        }
    }

    static void mergeSort(int arr[]) {
        int max = arr[0];
            for (int i : arr)
                if (max < i) max = i;

        mergeSort(arr, 0, arr.length - 1, max + 1);
    }
    public static void main(String[] args) {
        int arr[] = {4, 5, 6, 2, 38, 0, 3, 2, 8};
        mergeSort(arr);

        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}
